airbyte_cdk/sources/declarative/retrievers/async_retriever.py

```diff
@@ -116,7 +93,7 @@ def read_records(
     ) -> Iterable[StreamData]:
         stream_state: StreamState = self._get_stream_state()
         partition: AsyncPartition = self._validate_and_get_stream_slice_partition(stream_slice)
-        records: Iterable[Mapping[str, Any]] = self._job_orchestrator.fetch_records(partition)
+        records: Iterable[Mapping[str, Any]] = self.stream_slicer.fetch_records(partition)
```

Comment:
This line was surprising to me. I don't know why but the fact that we can retrieve records from a stream_slicer hurts my brain. I'm not strongly oppose to it though and maybe this is just a naming thing? I think I understand the challenge around the instantiation of the JobOrchestrator though and I'm out of naming for this one haha

Reply:
yeah so i really went back and forth on this one a lot yesterday. I agree that it feels a bit wrong that the new stream slicer is also providing the ability to fetch records since it's not in the interface and effectively this only works because the `AsyncRetriever` explicitly types `AsyncJobPartitionRouter` instead of the generic `StreamSlicer` interface.

The original way I had this was that I exposed a `AsyncJobPartitionRouter.job_orchestrator` property method, and kept old method:
```
def _job_orchestrator(self) -> AsyncJobOrchestrator:
    if not self.stream_slicer.job_orchestrator:
        raise AirbyteTracedException(
            message="Invalid state within AsyncJobRetriever. Please contact Airbyte Support",
            internal_message="AsyncPartitionRepository is expected to be accessed only after `stream_slices`",
            failure_type=FailureType.system_error,
        )

    return self.stream_slicer.job_orchestrator
```

So either way was left with a bit of an unfortunate side effect of relying on methods that only exist on `AsyncJobPartitionRouter` and not `StreamSlicer`. I'm not opposed to switching it back to this if you feel strongly about not intermixing `fetch_records()` onto the `AsyncJobPartitionRouter`